README.md
<div align="center">

# Docker Kava

## NOTICE: No Longer Maintained

<a href="https://github.com/Kava-Labs/"><img src="https://raw.githubusercontent.com/RyanHendricks/docker-kava/master/.github/kava.svg" alt="drawing" width="100"/></a>

[![Docker Pulls](https://img.shields.io/docker/pulls/ryanhendricks/docker-kava.svg?logo=docker&logoColor=white)](https://hub.docker.com/r/ryanhendricks/docker-kava) [![MicroBadger Layers (latest)](https://img.shields.io/microbadger/layers/ryanhendricks/docker-kava/latest.svg?logo=docker&logoColor=white)](https://microbadger.com/images/ryanhendricks/docker-kava) [![MicroBadger Version](https://images.microbadger.com/badges/version/ryanhendricks/docker-kava.svg)](https://microbadger.com/images/ryanhendricks/docker-kava) [![CircleCI (all branches)](https://img.shields.io/circleci/build/github/RyanHendricks/docker-kava/master?logo=circleci&token=5a11bae20d4346a49ee28e97d89c186f711a407c)](https://img.shields.io/circleci/build/github/RyanHendricks/docker-kava/master?logo=circleci&token=5a11bae20d4346a49ee28e97d89c186f711a407c) [![Codacy grade](https://img.shields.io/codacy/grade/738a34f90f1641ea9321abb27dd981f5.svg?logo=codacy)](https://www.codacy.com?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=RyanHendricks/docker-kava&amp;utm_campaign=Badge_Grade)

</div>

<!-- @import "[TOC]" {cmd="toc" depthFrom=1 depthTo=2 orderedList=false} -->

<!-- code_chunk_output -->

- [Docker Kava](#docker-kava)
  - [NOTICE: No Longer Maintained](#notice-no-longer-maintained)
  - [Prerequisites](#prerequisites)
  - [Quick Start](#quick-start)
  - [Configuration](#configuration)
    - [Config.toml Parameters](#configtoml-parameters)
    - [Environment Variables](#environment-variables)
    - [Bootstrapping](#bootstrapping)
  - [Build](#build)
  - [Run](#run)
  - [Supervisor](#supervisor)
    - [Kvd](#kvd)
    - [Kvcli Rest-Server](#kvcli-rest-server)
  - [Contributing](#contributing)
    - [Contributors](#contributors)
  - [License](#license)

<!-- /code_chunk_output -->

## Prerequisites

- Docker
- Docker-Compose (optional)

## Quick Start

The image can be run without any configuration and defaults to mainnet (currently kava-3)

```bash
docker  run --rm -it -P ryanhendricks/docker-kava:latest
```

## Configuration

### Config.toml Parameters

- The config.toml is created dynamically when starting the container.
- All parameters specified in the standard config.toml file can be set using environmental variables with the same name as the config parameter but in all caps. See [entrypoint script](./scripts/entrypoint.sh) for more details.
- If left unset the default values will be used.
- Parameters can be set directly by modifying the config.toml portion of ./scripts/entrypoint.sh if you are cloning and building the image yourself.

### Environment Variables

You can set ENV variables either in a docker-compose file or in the docker run command if running the container directly. If left unchanged they will default to the standard value except for the following:

- MONIKER
  - defaults to "nonamenode"
- CHAIN_ID
  - defaults to kava-4
- GENESIS_URL
  - defaults to kava-3 github [genesis file url](https://raw.githubusercontent.com/Kava-Labs/launch/master/kava-4/genesis.json)

Some examples:

```bash
MONIKER=nonamenode
CHAIN_ID=kava-3
BOOTSTRAP=TRUE
PROMETHEUS=true
GENESIS_URL=https://raw.githubusercontent.com/Kava-Labs/launch/master/kava-4/genesis.json
KVD_HOME=/.kvd
```

Running multiple nodes with the same IP address is possible using different ports. This is easily accomplished with the following ENV variabes.

```bash
PROMETHEUS_LISTEN_ADDR=46660
MAX_OPEN_CONNECTIONS=10
LCD_PORT=4317
RPC_PORT=46657
P2P_PORT=46656
PROXY_APP_PORT=46658
```

### Bootstrapping

  The entrypoint script allows for the node to be bootstrapped upon creation. This drastically reduces the time required to fully sync with the network. The start up time is increased relative to the amount of time it takes the host machine to download the archived data and extract it into the data directory.

  Set the ENV variable ```BOOTSTRAP=TRUE``` to enable bootstrapping.

## Build

The following command will build the image.

```bash
docker build --rm -f Dockerfile -t docker-kava:latest .
```

## Run

```sh
docker-compose up -d --build
```

## Supervisor

The image uses Supervisor to run both kvd and kvcli simultaneously (in the same container) at container runtime. Supervisor also restarts either process should it fail for some reason.

### Kvd

After starting the container you can check the status here: [http://0.0.0.0:26657/status](http://0.0.0.0:26657/status).

or from the terminal

```bash
curl -X GET \
  http://127.0.0.1:26657/status? \
  -H 'cache-control: no-cache'
```

### Kvcli Rest-Server

Supervisor starts the rest-server with the following command:

```bash
kvcli rest-server --trust-node --cors * --home $KVD_HOME --laddr tcp://0.0.0.0:1317

```

You can verify that the rest-server is running using the following example

```bash
curl -X GET \
  http://0.0.0.0:1317/blocks/latest \
  -H 'cache-control: no-cache'
```

## Contributing

[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?)](http://makeapullrequest.com)

### Contributors

[![Keybase PGP](https://img.shields.io/keybase/pgp/ryanhendricks.svg?label=keybase&logo=keybase&logoColor=white)](https://keybase.io/ryanhendricks)

## License

![GitHub](https://img.shields.io/github/license/ryanhendricks/docker-kava.svg)


.github/CHANGELOG.md
# Changelog

## [v0.11.0-kava-4](https://github.com/RyanHendricks/docker-kava/tree/v0.11.0-kava-4) (2020-10-22)

[Full Changelog](https://github.com/RyanHendricks/docker-kava/compare/v0.10.0-kava-3...v0.11.0-kava-4)

**Updates:**

- Kava-4 upgrade [\#2](https://github.com/RyanHendricks/docker-kava/pull/2)

## [v0.10.0-kava-3](https://github.com/RyanHendricks/docker-kava/tree/v0.10.0-kava-3) (2020-10-02)

[Full Changelog](https://github.com/RyanHendricks/docker-kava/compare/v1.0.0-rc1...v0.10.0-kava-3)

**Merged pull requests:**

- v1.0.0-rc2 [\#1](https://github.com/RyanHendricks/docker-kava/pull/1)

## [v1.0.0-rc1](https://github.com/RyanHendricks/docker-kava/tree/v1.0.0-rc1) (2020-02-01)

[Full Changelog](https://github.com/RyanHendricks/docker-kava/compare/v0.0.1...v1.0.0-rc1)

## [v0.0.1](https://github.com/RyanHendricks/docker-kava/tree/v0.0.1) (2020-01-05)

[Full Changelog](https://github.com/RyanHendricks/docker-kava/compare/473cd72e0314b521f9321b36a6b772aca680647c...v0.0.1)



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


scripts/entrypoint.sh
#!/bin/sh


# exit script on any error
set -e
echo "setting up initial configurations"

if [ ! -f "$KVD_HOME/config/config.toml" ];
then

  kvd init ${MONIKER:-nonamenode} --home=${KVD_HOME:-/.kvd} --chain-id=${CHAIN_ID:-kava-4}

  cd $KVD_HOME/config

  echo "removing autogenerated genesis and config files"
  rm genesis.json
  rm config.toml
  rm app.toml


  echo "downloading genesis.json"

  if [ ! -z "$GENESIS_URL" ]; then
      wget $GENESIS_URL
  else
      wget https://raw.githubusercontent.com/Kava-Labs/launch/master/kava-4/genesis.json
  fi

echo "generating config.toml"

cat > config.toml << EOF
proxy_app = "${PROXY_APP:-tcp://0.0.0.0}:${PROXY_APP_PORT:-26658}"
moniker = "${MONIKER:-moniker}"
fast_sync = ${FAST_SYNC:-true}
db_backend = "${DB_BACKEND:-goleveldb}"
db_dir = "${DB_DIR:-data}"
log_level = "${LOG_LEVEL:-main:info,state:info,*:error}"
log_format = "${LOG_FORMAT:-plain}"
genesis_file = "${GENESIS_FILE:-config/genesis.json}"
priv_validator_key_file = "${PRIV_VALIDATOR_KEY_FILE:-config/priv_validator_key.json}"
priv_validator_state_file = "${PRIV_VALIDATOR_KEY_FILE:-data/priv_validator_state.json}"
priv_validator_laddr = "${PRIV_VALIDATOR_LADDR:-}"
node_key_file = "${NODE_KEY_FILE:-config/node_key.json}"
abci = "${ABCI:-socket}"
prof_laddr = "${PROF_LADDR:-localhost:6060}"
filter_peers = ${FILTER_PEERS:-false}


[rpc]
laddr = "${RPC_LADDR:-tcp://0.0.0.0}:${RPC_PORT:-26657}"
cors_allowed_origins = ["${CORS_ALLOWED_ORIGINS:-*}"]
cors_allowed_methods = ["HEAD", "GET", "POST"]
cors_allowed_headers = ["Origin", "Accept", "Content-Type", "X-Requested-With", "X-Server-Time"]
grpc_laddr = "${GRPC_LADDR:-}"
grpc_max_open_connections = ${GRPC_MAX_OPEN_CONNECTIONS:-900}
unsafe = ${UNSAFE:-false}
max_open_connections = ${RPC_MAX_OPEN_CONNECTIONS:-900}
max_subscription_clients = ${MAX_SUBSCRIPTION_CLIENTS:-100}
max_subscriptions_per_client = ${MAX_SUBSCRIPTIONS_PER_CLIENT:-5}
timeout_broadcast_tx_commit = "${TIMEOUT_BROADCAST_TX_COMMIT:-10s}"
max_body_bytes = ${MAX_SIZE_REQUEST_BODY:-1000000}
max_header_bytes = ${MAX_SIZE_REQUEST_HEADER:-1048576}
tls_cert_file = "${TLS_CERT_FILE:-}"
tls_key_file = "${TLS_KEY_FILE:-}"


[p2p]
laddr = "${P2P_LADDR:-tcp://0.0.0.0}:${P2P_PORT:-26656}"
external_address = "${EXTERNAL_ADDRESS:-}"
seeds = "${SEEDS:-}"
persistent_peers = "${PERSISTENT_PEERS:-}"
upnp = ${UPNP:-false}
addr_book_file = "${ADDR_BOOK_FILE:-config/addrbook.json}"
addr_book_strict = ${ADDR_BOOK_STRICT:-true}
max_num_inbound_peers = ${MAX_NUM_INBOUND_PEERS:-40}
max_num_outbound_peers = ${MAX_NUM_OUTBOUND_PEERS:-10}
unconditional_peer_ids = "${UNCONDITIONAL_PEER_IDS:-}"
persistent_peers_max_dial_period = "${PERSISTENT_PEERS_MAX_DIAL_PERIOD:-0s}"
flush_throttle_timeout = "${FLUSH_THROTTLE_TIMEOUT:-100ms}"
max_packet_msg_payload_size = ${MAX_PACKET_MSG_PAYLOAD_SIZE:-1024}
send_rate = ${SEND_RATE:-5120000}
recv_rate = ${RECV_RATE:-5120000}
pex = ${PEX:-true}
seed_mode = ${SEED_MODE:-false}
private_peer_ids = "${PRIVATE_PEER_IDS:-}"
allow_duplicate_ip = ${ALLOW_DUPLICATE_IP:-false}
handshake_timeout = "${HANDSHAKE_TIMEOUT:-20s}"
dial_timeout = "${DIAL_TIMEOUT:-3s}"

[mempool]
recheck = ${RECHECK:-true}
broadcast = ${BROADCAST:-true}
wal_dir = "${WAL_DIR:-}"
size = ${SIZE_OF_MEMPOOL:-5000}
max_txs_bytes = ${MAX_TXS_BYTES:-1073741824}
cache_size = ${CACHE_SIZE:-10000}
max_tx_bytes = ${MAX_TX_BYTES:-1048576}

[fastsync]
version = "${FAST_SYNC_VERSION:-v0}"

[consensus]
wal_file = "${WAL_FILE:-data/cs.wal/wal}"
timeout_propose = "${TIMEOUT_PROPOSE:-3s}"
timeout_propose_delta = "${TIMEOUT_PROPOSE_DELTA:-500ms}"
timeout_prevote = "${TIMEOUT_PREVOTE:-1s}"
timeout_prevote_delta = "${TIMEOUT_PREVOTE_DELTA:-500ms}"
timeout_precommit = "${TIMEOUT_PRECOMMIT:-1s}"
timeout_precommit_delta = "${TIMEOUT_PRECOMMIT_DELTA:-500ms}"
timeout_commit = "${TIMEOUT_COMMIT:-5s}"
skip_timeout_commit = ${SKIP_TIMEOUT_COMMIT:-false}
create_empty_blocks = ${CREATE_EMPTY_BLOCKS:-true}
create_empty_blocks_interval = "${CREATE_EMPTY_BLOCKS_INTERVAL:-0s}"
peer_gossip_sleep_duration = "${PEER_GOSSIP_SLEEP_DURATION:-100ms}"
peer_query_maj23_sleep_duration = "${PEER_QUERY_MAJ23_SLEEP_DURATION:-2s}"

[tx_index]
indexer = "${INDEXER_SELECTION:-kv}"
# index_keys = "${INDEX_TAGS:-tx.height,tx.hash,tx.events,tx.data,tx.info}"
index_all_keys = ${INDEX_ALL_TAGS:-true}

[instrumentation]
prometheus = ${PROMETHEUS:-false}
prometheus_listen_addr = ":${PROMETHEUS_LISTEN_ADDR:-26660}"
max_open_connections = ${MAX_OPEN_CONNECTIONS:-3}
namespace = "${NAMESPACE:-tendermint}"
EOF


echo "generating app.toml"

cat > app.toml << EOF
minimum-gas-prices = "${MINIMUM_GAS_PRICES:-}"
halt-height = ${HALT_HEIGHT:-0}
halt-time = ${HALT_TIME:-0}
inter-block-cache = ${INTER_BLOCK_CACHE:-true}
pruning = "${PRUNING:-nothing}"
EOF



cd $KVD_HOME

  if [ "$BOOTSTRAP" == "TRUE" ]; then
    wget https://get.quicksync.io/kava-3-archive.20200929.0105.tar.lz4
    lz4 -d -v --rm kava-3-archive.20200929.0105.tar.lz4 | tar xf -
  fi

fi


echo "updating supervisor config files"
rm /etc/supervisor/conf.d/supervisor-kvcli.conf
rm /etc/supervisor/conf.d/supervisor-kvd.conf

cd /etc/supervisor/conf.d/

cat > supervisor-kvcli.conf << EOF
[program:kvcli]
command=kvcli rest-server --laddr tcp://0.0.0.0:${LCD_PORT:-1317} --home=${KVD_HOME:-/.kvd} --chain-id=${CHAIN_ID:-kava-4} --trust-node --node=${RPC_LADDR:-tcp://0.0.0.0}:${RPC_PORT:-26657}
redirect_stderr=false
autostart=true
autorestart=unexpected
startsecs=10
startretries=5
stdout_logfile=/stdout-kvcli.txt
stdout_logfile_maxbytes=10MB
stdout_logfile_backups=10
stdout_capture_maxbytes=10MB
stdout_events_enabled=false
stderr_logfile=/stderr-kvcli.txt
stderr_logfile_maxbytes=10MB
stderr_logfile_backups=10
stderr_capture_maxbytes=10MB
stderr_events_enabled=false
EOF

cat > supervisor-kvd.conf << EOF
[program:kvd]
command=kvd start --home=${KVD_HOME:-/.kvd}
redirect_stderr=false
autostart=true
autorestart=unexpected
startsecs=10
startretries=5
stdout_logfile=/stdout-kvd.txt
stdout_logfile_maxbytes=10MB
stdout_logfile_backups=10
stdout_capture_maxbytes=10MB
stdout_events_enabled=false
stderr_logfile=/stderr-kvd.txt
stderr_logfile_maxbytes=10MB
stderr_logfile_backups=10
stderr_capture_maxbytes=10MB
stderr_events_enabled=false
EOF


echo "configuration complete  ---- starting..."
exec supervisord --nodaemon --configuration /etc/supervisor/supervisord.conf
